File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/cassandra/extension/CassandraTypesExtension.java
package ru.tinkoff.kora.database.annotation.processor.cassandra.extension;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.ParameterizedTypeName;
import com.palantir.javapoet.TypeName;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.GenericTypeResolver;
import ru.tinkoff.kora.annotation.processor.common.NameUtils;
import ru.tinkoff.kora.database.annotation.processor.cassandra.CassandraTypes;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.type.TypeVariable;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Map;

public class CassandraTypesExtension implements KoraExtension {

    private final Types types;
    private final Elements elements;
    //    private final CassandraEntityGenerator generator;

    public CassandraTypesExtension(ProcessingEnvironment env) {
        this.types = env.getTypeUtils();
        this.elements = env.getElementUtils();
    }

    @Nullable
    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, @Nullable String tag) {
        if (tag != null) {
            return null;
        }
        var typeName = TypeName.get(typeMirror).withoutAnnotations();
        if (!(typeName instanceof ParameterizedTypeName ptn) || !(typeMirror instanceof DeclaredType dt)) {
            return null;
        }
        if (ptn.rawType().equals(CassandraTypes.RESULT_SET_MAPPER)) {
            return this.generateResultSetMapper(roundEnvironment, dt);
        }
        if (ptn.rawType().equals(CassandraTypes.ASYNC_RESULT_SET_MAPPER)) {
            return this.generateAsyncResultSetMapper(roundEnvironment, dt);
        }
        if (ptn.rawType().equals(CassandraTypes.ROW_MAPPER)) {
            return this.generateResultRowMapper(roundEnvironment, dt);
        }
        if (ptn.rawType().equals(CassandraTypes.PARAMETER_COLUMN_MAPPER)) {
            return this.generateParameterColumnMapper(roundEnvironment, dt);
        }
        if (ptn.rawType().equals(CassandraTypes.RESULT_COLUMN_MAPPER)) {
            return this.generateRowColumnMapper(roundEnvironment, dt);
        }
        return null;
    }

    private KoraExtensionDependencyGenerator generateRowColumnMapper(RoundEnvironment roundEnvironment, DeclaredType dt) {
        var entityType = dt.getTypeArguments().get(0);
        var element = (TypeElement) this.types.asElement(entityType);
        if (AnnotationUtils.findAnnotation(element, CassandraTypes.UDT_ANNOTATION) != null) {
            return KoraExtensionDependencyGenerator.generatedFrom(elements, element, CassandraTypes.RESULT_COLUMN_MAPPER);
        }
        if (element.getQualifiedName().contentEquals("java.util.List")) {
            entityType = ((DeclaredType) entityType).getTypeArguments().get(0);
            element = (TypeElement) this.types.asElement(entityType);
            if (AnnotationUtils.findAnnotation(element, CassandraTypes.UDT_ANNOTATION) != null) {
                return KoraExtensionDependencyGenerator.generatedFrom(elements, element, "List_CassandraRowColumnMapper");
            }
        }
        return null;
    }

    private KoraExtensionDependencyGenerator generateParameterColumnMapper(RoundEnvironment roundEnvironment, DeclaredType dt) {
        var entityType = dt.getTypeArguments().get(0);
        var element = (TypeElement) this.types.asElement(entityType);
        if (AnnotationUtils.findAnnotation(element, CassandraTypes.UDT_ANNOTATION) != null) {
            return KoraExtensionDependencyGenerator.generatedFrom(elements, element, CassandraTypes.PARAMETER_COLUMN_MAPPER);
        }
        if (element.getQualifiedName().contentEquals("java.util.List")) {
            entityType = ((DeclaredType) entityType).getTypeArguments().get(0);
            element = (TypeElement) this.types.asElement(entityType);
            if (AnnotationUtils.findAnnotation(element, CassandraTypes.UDT_ANNOTATION) != null) {
                return KoraExtensionDependencyGenerator.generatedFrom(elements, element, "List_CassandraParameterColumnMapper");
            }
        }
        return null;
    }

    @Nullable
    private KoraExtensionDependencyGenerator generateResultRowMapper(RoundEnvironment roundEnvironment, DeclaredType typeMirror) {
        var rowType = typeMirror.getTypeArguments().get(0);
        var rowTypeElement = this.types.asElement(rowType);
        if (AnnotationUtils.isAnnotationPresent(rowTypeElement, CassandraTypes.CASSANDRA_ENTITY)) {
            return KoraExtensionDependencyGenerator.generatedFrom(elements, rowTypeElement, CassandraTypes.ROW_MAPPER);
        }
        return null;
    }

    @Nullable
    private KoraExtensionDependencyGenerator generateResultSetMapper(RoundEnvironment roundEnvironment, DeclaredType typeMirror) {
        //CassandraResultSetMapper<List<T>>
        var resultType = typeMirror.getTypeArguments().get(0);
        if (!(resultType instanceof DeclaredType dt)) {
            return null;
        }
        if (CommonUtils.isList(resultType)) {
            var tn = (ParameterizedTypeName) TypeName.get(resultType);
            var rowType = dt.getTypeArguments().get(0);
            return this.listResultSetMapper(typeMirror, tn, (DeclaredType) rowType);
        }
        var rowTypeElement = this.types.asElement(resultType);
        if (AnnotationUtils.isAnnotationPresent(rowTypeElement, CassandraTypes.CASSANDRA_ENTITY)) {
            return KoraExtensionDependencyGenerator.generatedFrom(elements, rowTypeElement, CassandraTypes.RESULT_SET_MAPPER);
        }
        return () -> {
            var singleResultSetMapper = findStaticMethod(CassandraTypes.RESULT_SET_MAPPER, "singleResultSetMapper");
            var tp = (TypeVariable) singleResultSetMapper.getTypeParameters().get(0).asType();
            var executableType = (ExecutableType) GenericTypeResolver.resolve(this.types, Map.of(tp, resultType), singleResultSetMapper.asType());
            return ExtensionResult.fromExecutable(singleResultSetMapper, executableType);
        };
    }

    @Nullable
    private KoraExtensionDependencyGenerator generateAsyncResultSetMapper(RoundEnvironment roundEnvironment, DeclaredType typeMirror) {
        //CassandraResultSetMapper<List<T>>
        var listType = typeMirror.getTypeArguments().get(0);
        if (!(listType instanceof DeclaredType dt)) {
            return null;
        }
        if (CommonUtils.isList(listType)) {
            return () -> {
                var tn = (ParameterizedTypeName) TypeName.get(listType);
                var rowType = dt.getTypeArguments().get(0);
                var singleResultSetMapper = findStaticMethod(CassandraTypes.ASYNC_RESULT_SET_MAPPER, "list");
                var tp = (TypeVariable) singleResultSetMapper.getTypeParameters().get(0).asType();
                var executableType = (ExecutableType) GenericTypeResolver.resolve(this.types, Map.of(tp, rowType), singleResultSetMapper.asType());
                return ExtensionResult.fromExecutable(singleResultSetMapper, executableType);
            };
        }
        return () -> {
            var singleResultSetMapper = findStaticMethod(CassandraTypes.ASYNC_RESULT_SET_MAPPER, "one");
            var tp = (TypeVariable) singleResultSetMapper.getTypeParameters().get(0).asType();
            var executableType = (ExecutableType) GenericTypeResolver.resolve(this.types, Map.of(tp, listType), singleResultSetMapper.asType());
            return ExtensionResult.fromExecutable(singleResultSetMapper, executableType);
        };
    }

    private KoraExtensionDependencyGenerator listResultSetMapper(DeclaredType typeMirror, ParameterizedTypeName listType, DeclaredType rowTypeMirror) {
        var rowTypeElement = this.types.asElement(rowTypeMirror);
        if (AnnotationUtils.isAnnotationPresent(rowTypeElement, CassandraTypes.CASSANDRA_ENTITY)) {
            return KoraExtensionDependencyGenerator.generatedFromWithName(elements, rowTypeElement, NameUtils.generatedType(rowTypeElement, "ListCassandraResultSetMapper"));
        }
        return () -> {
            var listResultSetMapper = findStaticMethod(CassandraTypes.RESULT_SET_MAPPER, "listResultSetMapper");
            var tp = (TypeVariable) listResultSetMapper.getTypeParameters().get(0).asType();
            var executableType = (ExecutableType) GenericTypeResolver.resolve(this.types, Map.of(tp, rowTypeMirror), listResultSetMapper.asType());
            return ExtensionResult.fromExecutable(listResultSetMapper, executableType);
        };
    }

    private ExecutableElement findStaticMethod(ClassName className, String methodName) {
        return this.elements.getTypeElement(className.canonicalName()).getEnclosedElements()
            .stream()
            .filter(e -> e.getKind() == ElementKind.METHOD && e.getModifiers().contains(Modifier.STATIC))
            .map(ExecutableElement.class::cast)
            .filter(m -> m.getSimpleName().contentEquals(methodName))
            .findFirst()
            .orElseThrow();
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/cassandra/extension/CassandraTypesExtension.kt
package ru.tinkoff.kora.database.symbol.processor.cassandra.extension

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.getFunctionDeclarationsByName
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.Variance
import ru.tinkoff.kora.database.symbol.processor.cassandra.CassandraEntityGenerator
import ru.tinkoff.kora.database.symbol.processor.cassandra.CassandraTypes
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.ksp.common.CommonClassNames.isList
import ru.tinkoff.kora.ksp.common.KspCommonUtils.parametrized

//CassandraRowMapper<T>
//CassandraResultSetMapper<List<T>>
class CassandraTypesExtension(val resolver: Resolver, val kspLogger: KSPLogger, val codeGenerator: CodeGenerator) : KoraExtension {
    private val entityGenerator = CassandraEntityGenerator(codeGenerator)

    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tag: String?): (() -> ExtensionResult)? {
        if (tag != null) return null
        if (type.declaration.qualifiedName?.asString()?.equals(CassandraTypes.resultSetMapper.canonicalName) == true) {
            return this.generateResultSetMapper(resolver, type)
        }
        if (type.declaration.qualifiedName?.asString()?.equals(CassandraTypes.asyncResultSetMapper.canonicalName) == true) {
            return this.generateAsyncResultSetMapper(resolver, type)
        }
        if (type.declaration.qualifiedName?.asString()?.equals(CassandraTypes.rowMapper.canonicalName) == true) {
            return this.generateRowMapper(resolver, type)
        }
        if (type.declaration.qualifiedName?.asString()?.equals(CassandraTypes.parameterColumnMapper.canonicalName) == true) {
            return this.generateParameterColumnMapper(resolver, type)
        }
        if (type.declaration.qualifiedName?.asString()?.equals(CassandraTypes.rowColumnMapper.canonicalName) == true) {
            return this.generateRowColumnMapper(resolver, type)
        }

        return null
    }

    private fun generateResultSetMapper(resolver: Resolver, rowSetKSType: KSType): (() -> ExtensionResult)? {
        val rowSetParam = rowSetKSType.arguments[0].type!!.resolve()
        if (rowSetParam.isMarkedNullable) {
            return null
        }
        if (!rowSetParam.isList()) {
            if (rowSetParam.declaration.isAnnotationPresent(CassandraTypes.entity)) {
                return generatedByProcessor(resolver, rowSetParam.declaration as KSClassDeclaration, CassandraTypes.resultSetMapper)
            }
            val resultSetMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.resultSetMapper.canonicalName)!!
            val rowMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.rowMapper.canonicalName)!!
            val resultSetMapperType = resultSetMapperDecl.asType(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(rowSetParam), Variance.INVARIANT)))
            val rowMapperType = rowMapperDecl.asType(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(rowSetParam), Variance.INVARIANT)))

            val functionDecl = resolver.getFunctionDeclarationsByName(CassandraTypes.resultSetMapper.canonicalName + ".singleResultSetMapper").first()
            val functionType = functionDecl.parametrized(resultSetMapperType, listOf(rowMapperType))
            return {
                ExtensionResult.fromExecutable(functionDecl, functionType)
            }
        }
        val rowType = rowSetParam.arguments[0]
        val rowResolvedType = rowType.type!!.resolve()
        if (rowResolvedType.declaration.isAnnotationPresent(CassandraTypes.entity)) {
            return generatedByProcessor(resolver, rowResolvedType.declaration as KSClassDeclaration, "ListCassandraResultSetMapper")
        }
        val resultSetMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.resultSetMapper.canonicalName)!!
        val rowMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.rowMapper.canonicalName)!!

        val resultSetMapperType = resultSetMapperDecl.asType(
            listOf(
                resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(rowSetParam), Variance.INVARIANT)
            )
        )
        val rowMapperType = rowMapperDecl.asType(
            listOf(
                resolver.getTypeArgument(rowSetParam.arguments[0].type!!, Variance.INVARIANT)
            )
        )

        val functionDecl = resolver.getFunctionDeclarationsByName(CassandraTypes.resultSetMapper.canonicalName + ".listResultSetMapper").first()
        val functionType = functionDecl.parametrized(resultSetMapperType, listOf(rowMapperType))
        return {
            ExtensionResult.fromExecutable(functionDecl, functionType)
        }
    }

    private fun generateRowMapper(resolver: Resolver, rowKSType: KSType): (() -> ExtensionResult)? {
        val rowType = rowKSType.arguments[0].type!!.resolve()
        if (rowType.declaration.isAnnotationPresent(CassandraTypes.entity)) {
            return generatedByProcessor(resolver, rowType.declaration as KSClassDeclaration, CassandraTypes.rowMapper)
        }
        return null
    }

    private fun generateAsyncResultSetMapper(resolver: Resolver, type: KSType): (() -> ExtensionResult)? {
        val resultType = type.arguments[0].type!!.resolve()
        if (resultType.isMarkedNullable) {
            return null
        }
        val resultClassDecl = resultType.declaration as KSClassDeclaration
        if (resultClassDecl.qualifiedName?.asString() == "kotlin.collections.List") {
            val rowType = resultType.arguments[0].type!!.resolve()
            if (rowType.isMarkedNullable) {
                return null
            }
            val resultSetMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.asyncResultSetMapper.canonicalName)!!
            val rowMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.rowMapper.canonicalName)!!

            val resultSetMapperType = resultSetMapperDecl.asType(
                listOf(
                    resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(resultType), Variance.INVARIANT)
                )
            )
            val rowMapperType = rowMapperDecl.asType(
                listOf(
                    resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(rowType), Variance.INVARIANT)
                )
            )

            val functionDecl = resolver.getFunctionDeclarationsByName(CassandraTypes.asyncResultSetMapper.canonicalName + ".list").first()
            val functionType = functionDecl.parametrized(resultSetMapperType, listOf(rowMapperType))
            return {
                ExtensionResult.fromExecutable(functionDecl, functionType)
            }
        } else {
            val resultSetMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.asyncResultSetMapper.canonicalName)!!
            val rowMapperDecl = resolver.getClassDeclarationByName(CassandraTypes.rowMapper.canonicalName)!!

            val resultSetMapperType = resultSetMapperDecl.asType(
                listOf(
                    resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(resultType), Variance.INVARIANT)
                )
            )
            val rowMapperType = rowMapperDecl.asType(
                listOf(
                    resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(resultType), Variance.INVARIANT)
                )
            )

            val functionDecl = resolver.getFunctionDeclarationsByName(CassandraTypes.asyncResultSetMapper.canonicalName + ".one").first()
            val functionType = functionDecl.parametrized(resultSetMapperType, listOf(rowMapperType))
            return {
                ExtensionResult.fromExecutable(functionDecl, functionType)
            }
        }
    }

    private fun generateParameterColumnMapper(resolver: Resolver, type: KSType): (() -> ExtensionResult)? {
        val entityType = type.arguments[0].type!!.resolve()
        if (entityType.isMarkedNullable) {
            return null
        }
        val ksClassDeclaration = entityType.declaration as KSClassDeclaration
        if (ksClassDeclaration.findAnnotation(CassandraTypes.udt) != null) {
            return generatedByProcessor(resolver, ksClassDeclaration, "CassandraParameterColumnMapper")
        }
        if (ksClassDeclaration.qualifiedName?.asString() == "kotlin.collections.List") {
            val t = entityType.arguments[0].type!!.resolve()
            if (t.isMarkedNullable) {
                return null
            }
            val listElementClassDeclaration = t.declaration as KSClassDeclaration
            if (listElementClassDeclaration.findAnnotation(CassandraTypes.udt) != null) {
                return generatedByProcessor(resolver, listElementClassDeclaration, "List_CassandraParameterColumnMapper")
            }
        }
        return null
    }

    private fun generateRowColumnMapper(resolver: Resolver, type: KSType): (() -> ExtensionResult)? {
        val entityType = type.arguments[0].type!!.resolve()
        if (entityType.isMarkedNullable) {
            return null
        }
        val ksClassDeclaration = entityType.declaration as KSClassDeclaration
        if (ksClassDeclaration.findAnnotation(CassandraTypes.udt) != null) {
            return generatedByProcessor(resolver, ksClassDeclaration, "CassandraRowColumnMapper")
        }
        if (ksClassDeclaration.qualifiedName?.asString() == "kotlin.collections.List") {
            val t = entityType.arguments[0].type!!.resolve()
            if (t.isMarkedNullable) {
                return null
            }
            val listKsClassDeclaration = t.declaration as KSClassDeclaration
            if (listKsClassDeclaration.findAnnotation(CassandraTypes.udt) != null) {
                return generatedByProcessor(resolver, listKsClassDeclaration, "List_CassandraRowColumnMapper")
            }
        }

        return null
    }

}
